File: Java/src/main/java/org/onlineoptimisation/problem/benchmark/DiscusFunction.java
package org.onlineoptimisation.problem.benchmark;

import org.armadillojava.Arma;
import org.armadillojava.Col;
import org.armadillojava.Mat;

public class DiscusFunction extends BenchmarkProblem {

  protected Mat _R;
  
  public DiscusFunction(int numberOfDimensions) {
    super(numberOfDimensions);
    
    _R = randomRotation();
  }

  @Override
  public double getObjectiveValueImplementation(Col parameter) {
    Col z = Arma.square(oscillate(_R.times(parameter)));
    return 999999 * z.at(1) + Arma.accu(z);
  }

}

File: src/optimisationProblem/benchmark/discusFunction.cpp
#include <hop_bits/optimisationProblem/benchmark/discusFunction.hpp>

namespace hop {
  double DiscusFunction::getObjectiveValueImplementation(const arma::Col<double>& parameter) const {
    arma::Col<double> z = arma::square(getOscillationTransformation(rotationR_ * (parameter - translation_)));
    return 1000000 * z.at(0) + arma::accu(z.subvec(1, z.n_elem - 1));
  }
}
